### DISCUSSION TITLE
[Re-opened] Interpolation scheme to be used in WAFS SIGWX Forecast

### DISCUSSION THREAD

**Discussion by blchoy** [wmo-im/iwxxm#263] [Resolution]
The document [_Representing WAFS SIGWX Data in BUFR_](https://www.icao.int/airnavigation/METP/MOGWAFS%20Reference%20Documents/Representing%20WAFS%20SIGWX%20Data%20in%20BUFR%20-%20V4%203%20Final.pdf) mentioned that _"An appropriate smoothing technique such as a cubic spline should be used to smooth the jet-stream curve plotted between the core points."_ This also applies to the horizontal boundary of the CAT and cloud areas.

In the IWXXM implementation of the WAFS SIGWX Forecast, it is anticipated to carry forward this idea in the encoding:
```
<!-- A Jet Stream Object -->
<iwxxm:feature>
    <iwxxm:MeteorologicalFeature gml:id="uuid.8c3d3467-e398-4627-ab23-f9e57c2215b5">
        <gml:identifier codeSpace="http://wafs/sigwxobj">eb324455-236c-46f8-b80d-c6d42c3a634c</gml:identifier>
        <iwxxm:phenomenon xlink:href="http://codes.wmo.int/49-2/MeteorologicalFeature/JETSTREAM"/>
        <iwxxm:phenomenonGeometry>
            <gml:Curve gml:id="uuid.7b5a197a-d9bf-4f3b-ba36-701fe019b6d2" srsDimension="2" axisLabels="Lat Long" srsName="http://www.opengis.net/def/ers/EPSG/0/4326">
                <gml:segments>
                    <gml:CubicSpline>
                        <gml:posList>35.4 -104.7 36.1 -101.6 36.6 -99.6 36.9 -97.7 37.4 -94.9 38.3 -92.3 39.2 -90.3 40.7 -88.1 41.5 -86.3 42.0 -82.7 41.9 -78.9</gml:posList>
                        <gml:vectorAtStart>1.0 2.0</gml:vectorAtStart>
                        <gml:vectorAtEnd>2.0 3.0</gml:vectorAtEnd>
                    </gml:CubicSpline>
                </gml:segments>
            </gml:Curve>
        </iwxxm:phenomenonGeometry>
        ...
```

@borisburger in [his e-mail](https://github.com/wmo-im/iwxxm/files/6471240/Re_.SIGWX.forecast.-.gml_CubicSpline.pdf) indicated difficulties in proper implementation of cubic splines and propose the use of Bezier.  @GAndrsn after studying Boris' suggestion had identify [some other issues](https://github.com/wmo-im/iwxxm/files/6471326/Re_.SIGWX.forecast.-.gml_CubicSpline_reply.pdf) with Bezier.

Summarizing the discussion so far, I think the key issues on this topic is as follow:
1. For WAFCs a smooth interpolation scheme is needed for both the boundaries of areas and the jet core.  it is essential for the interpolated curve to pass through the points indicated in the gml:posList, especially in the latter case as the direction of the jet stream is derived from the interpolated curve.
2. For software developers it is difficult, if not impossible, to properly implement cubic spline interpolation, especially if the application requires realtime adjustment to the curves like resizing of the polygon.
3. The CRS involved (i.e. EPSG:4326) further complicates this issue.

I hope I have given a correct account to the issue.  Graeme and Boris feel free to express your views.  Others are most welcome to join the discussion.

**Comment by borisburger** [comment]
I have read the response from Graeme and I completely agree with the view that forecasters do not ever think about the particular interpolation technique used by the underlying software tool. However an advantage of using Bezier cubic splines is that the software will directly encode what user had on screen without any room for interpretation - so the recipients of the data will draw the same shape the forecaster saw on screen.

This is because all major drawing APIs on Linux & Windows like Microsoft GDI, GDI+, Direct 2D/Direct Draw, Cairo, Skia, OpenGL etc. support cubic **Bezier** splines out of the box. Vendors then have a very easy job of plotting the curve with a single API call.

My main concern with the proposed **gml:CubicSpline** is not the difficulty in implementation - this kind of cubic spline is uncommon, but not overly difficult to implement. **Why do I say that it is uncommon?**

The real issue is with the underlying **GM_CubicSpline** in **ISO 19107 Geographic Information - Spatial Schema** where a very unfortunate name/terminology was chosen - using the general term "cubic spline" to actually denote a so called "C2 continuous natural cubic spline" (a standard textbook name). Quoting from ISO 19107 Geographic Information - Spatial Schema, section **7.13.6 Interface CubicSpline**:

> the curve must be C2, i.e. have a continuous first and second derivative at all points ... The control parameters record must contain derivativeAtStartvectorAtStart, and derivativeAtEndvectorAtEnd which are the tangent vectors at controlPoint[1] and controlPoint [n] ...

The above is a very particular and unique type of C2 continuous cubic splines (C2 = continuous 1st and 2nd derivatives). The general term "cubic spline" would normally denote a whole class of cardinal, Catmull-Rom, Kochanek-Bartels, Hermitean and other types of polynomial cubic splines. What is denoted by gml:CubicSpline is very confusingly not this whole class, but a very specific C2 continuous type. Please note in the quote from ISO 19107 that even with gml:CubicSpline you still need to specify derivatives at the curve endpoints! (but it is true you do not need to specify the derivatives at the inner control points where the segments join, because these can be fully determined from the C2 continuity requirement - but for the outer endpoints it is still necessary).

A good overview article is https://en.wikipedia.org/wiki/Spline_(mathematics) which mentions many cubic spline types and also explains the specific **natural** C2 cubic splines as used in GML. Also https://en.wikipedia.org/wiki/Cubic_Hermite_spline

The issue I see with this is that not everyone studied computational geometry or computer graphics. I am afraid some vendors due to the misleading (overly broad) term  **gml:CubicSpline** will not realise, that in GML (based on ISO 19107 GM_CubicSpline) this is a very special type of "C2 continuous natural cubic splines" and instead they will just use the same spline algorithm they have been using for the past 20 years - not displaying what was originally intended.

A brief summary about gml:CubicSpline (GM_CubicSpline):

1. The curve passes through all its control points and is smooth in mathematical sense = both 1st and 2nd derivative are continous along the entirety of the curve
2. Derivatives at outer endpoints of open curves (e.g. jetstreams) still need to be provided for every gml:CubicSpline
3. Implementations are unfortunately not readily available in major drawing APIs.
4. Implementation requires solving a 3-diagonal system of linear equations to figure out the 1st and 2nd derivatives at control points, and then convert to a Bezier to satisfy the graphics APIs that expect Beziers. The number of equations in the system is equal to the number of spline control points (100 control points leads to a system of 100 equations with 100 unknowns). It is however a textbook algorithm which can be found online and is taught in geometry/CG classes at universities, however it is not clear if people will realise this, because:
5. I have never seen C2 continuous natural cubic splines offered in graphics software like Autodesk AutoCAD, Fusion 360, or vector drawing software like Inkscape. I do not own a license for Adobe Illustrator, so I do not know if this type of C2 spline is offered there. The most commonly quoted reason for graphics software not offering C2 natural cubics is their "lack of local control" = if you move a single control point, you need to recalculate the entire 3-diagonal system of equations from scratch and the entirety of the curve changes striving to satisfy the C2 continuity. This property makes the C2 natural cubic splines a bit too unpredictable for graphics artists.

I am not fully opposed to using gml:CubicSpline. It is an intriguing concept and in GML besides gml:Bezier and gml:LinearRing there is no other reasonable choice. I fully recognize the advantage of not having to specify the additional control points like in Bezier to determine derivatives (however at outer endpoints they still need to be provided!). And the compact representation offered by gml:CubicSpline where the spline passes through all the specified control points is also a useful property.

I am just saying that the choice of gml:CubicSpline has a potential downside, that implementors will get confused by the overly generic "cubic spline" term and fall back on their existing algorithms, instead of going the more difficult route of solving the 3-diagonal system of equations for each curve. For those that will take the extra step this will make the software less efficient because of the additional computational complexity as opposed to Bezier or LinearRing. But I guess the overall impact of the extra math will not be so large considering all the other computations that need to be performed to plot the data on screen.

Until now I was trying to represent the broader community of software vendors, but I would appreciate if members of other companies wrote their point of view. Specifically for IBL, for manual drawing of SIGWX-like charts by the forecaster (in our drawing tools at least) we will still have to fall back on just C1 continuous cubic splines. Our company is currently using Cardinal cubic splines for SIGWX charts (https://en.wikipedia.org/wiki/Cubic_Hermite_spline) with c=0 tension parameter and a specific knot sequence. It is a simpler type of a cubic spline which makes approximation easier (approximation of jets or clouds drawn freehand on a tablet, or approximation of isocontours based on e.g. NWP MSLP pressure or geopotential). For the C2 natural cubics we do not know how to implement such numerical approximation algorithms that fit a 2D freehand curve. So if SIGWX switches over to C2 splines we will likely need to support also C2 continuous gml:CubicSpline and then approximate it with the Cardinal C1 cubic spline if users will want to continue tweaking the SIGWX charts for their area of interest (we have many users who base their mid-level SIGWX charts on the WAFC SIGWX data by importing the jets/clouds/CATs into graphical editor and moving control points around).

Kind regards,
Boris

**Comment by blchoy** [comment]
I see.  That means we need to skip using gml:CubicSpline and define our own representation of a curve with points along it and interpolation scheme like C1 cubic spline?

**Comment by GAndrsn** [comment]
I've had a read through all of the info and attachments in the discussion so far, and I think I have a broad understanding of the issues. I never studied geometry in depth, however, so I'm only gradually wrapping my head around Bezier curves and all the different types of splines.

I can see the advantages of curves defined as Bezier curves being totally unambiguous, whereas points described simply as defining a "spline" could mean a variety of curves. The only reason I have used cubic splines in the way I have so far is that this is what I inherited when I took on the SigWx code, and I didn't have much of an understanding of the complexities of curves at the time (and still don't fully, to be honest!).

I don't object to trying to implement using Bezier curves. I googled how to convert splines to Bezier curves earlier, and funnily enough came across the same link in Jan's email which I didn't look at until later. I can have a go and seeing what effect implementing this has, although it may take me a little while because, as I say, my knowledge of curve geometry is currently relatively limited.

One of my concerns would be making sure that all of the eventual users of the data are absolutely clear that going from manually-drawn objects in BUFR data to algorithmically-drawn objects in IWXXM that they can't just use their old spline functions, as the Bezier control points will presumably be outside of the hazard area perimeter, rather than on the perimeter itself. I'll discuss this with my colleagues who deal with the users of this sort of data.

**Reply by borisburger** [reply]
Graeme if you have a cubic spline with control points it passes through and you have the 1st order derivatives at the control points, then to convert any of the spline segments to a Bezier:

- Assume A and B are the endpoints of a segment
- Assume dA and dB are the 1st order derivatives of the spline at points A and B respectively (more or less dA and dB are 2D vectors tangential to the curve at A and B)

Then the Bezier control points P0, P1, P2, P3 for the single segment are:

- P0 = A
- P1 = A + dA/3
- P2 = B - dB/3
- P3 = B

Please see https://en.wikipedia.org/wiki/B%C3%A9zier_curve#Cubic_B%C3%A9zier_curves for more details, the page explains how derivatives of Beziers are calculated and the above is just a simple consequence of the equations.

In reverse if you need to calculate derivatives of a cubic Bezier segment:
dA = 3*(P1 - P0)
dB = 3*(P3 - P2)

**Reply by borisburger** [reply]
That being said, I feel kind of bad for pushing my views on the types of splines, and would really appreciate if some other people who will be implementing the display of IWXXM SIGWX data expressed their opinion, because at the end of the day I am also not 100% sure which route is the most reasonable to take in this case. @blchoy do you think there might be someone at e.g. DWD or MeteoFrance who would be willing to participate in this discussion?

**Reply by blchoy** [reply]
> @blchoy do you think there might be someone at e.g. DWD or MeteoFrance who would be willing to participate in this discussion?

We have team members from DWD (@dzinkhan) and Meteo France (@yg31).  Could you also help to promulgate this post to the MET service providers you know?  @mgoberfield it would be nice if you could also bring this to Matt's attention and see if there is also any views from WAFC Washington.

**Reply by GAndrsn** [reply]
@borisburger I've had a go at looking at this today but I think I'm just giving myself a headache. Understanding how to convert B-splines to Bezier curves using Python seems to require a much deeper understanding of the foundation mathematics than I have. To generate SigWx Objects I've been using the [SciPy splprep function](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.splprep.html), but as far as I can figure out it doesn't return values that I can determine derivatives from which I can use to convert the curve to a Bezier curve.

**Reply by GAndrsn** [reply]
I'm not giving up on working this out. Question: to represent SigWx objects in IWXXM as Bezier curves, is the idea to represent the shapes as very high order Bezier curves, or split the shape into subsegments between each pair of paremeter points, and add a couple of control points between these pairs to define a 3rd order Bezier curve between each pair of points around the shape perimeter?

**Reply by borisburger** [reply]
@GAndrsn I am not an expert on B-Splines, I know they are supported by GML too as gml:BSpline though. They are a more advanced/powerful spline technique. For plotting cubic B-Splines using common graphics APIs or storing into SVG (which supports just Beziers) they usually have to be converted to Beziers using a so called "Boehm algorithm" of "knot insertion". This is mentioned for example in this Stack Overflow thread: https://stackoverflow.com/questions/56859792/how-to-create-bezier-curves-from-b-splines-in-sympy (the last comment also tries to implement the conversion from scipy.interpolate.BSpline to a Bezier).

B-Splines are also a bit "unusual" because the curve does not pass through any of its control points (except for the very first and very last, but if you have 30 points inbetween, they just control the curvature somehow but the curve does not pass through them). See e.g. the first image on top of https://en.wikipedia.org/wiki/B-spline

While I do understand the C2 natural gml:CubicSpline and gml:Bezier, the gml:BSpline is a bit too over my head... I find it a bit strange that SciPy does not offer a function for conversion of BSpline to Bezier, because that Boehm knot insertion technique is a textbook method for doing just that (I remember this is teached in computer graphics courses at universities, but I found the BSplines a bit overwhelming - which could be just my own limitations, because I know they are sometimes offered in drawing applications and CAD tools).

The idea is to end up with a cubic spline, i.e. spline consisting of many segments which are cubic polynomials, exactly as you say. If these segments are cubic polynomials in the Bezier representation, it is more straightforward for people to just pass the control points into a drawing API.

What do you have on input to SciPy splprep? A list of discrete 2D points that should be fitted with a cubic B-spline?

**Reply by GAndrsn** [reply]
I can sympathise with BSplines feeling over your head! I've looked through the links provided but haven't found anything I can understand enough to use yet.

The input to splprep is the 2D coordinate points of input polygons, with the smoothing argument set to zero which makes the output curve pass through the input points.

**Reply by GAndrsn** [reply]
I've had a go at investigating the example code for the last comment in the Stack Overflow link you provided, but haven't gleaned much from it yet. It uses the aggdraw package, for which there is very little information online, so I can't replicated what it's doing fully. It seems to be adding points to the input point array, but I'm not sure how this is useful yet.

**Reply by GAndrsn** [reply]
I've been working my way though notes on splines for the University of Oslo, and I think it's helping me understand things. I think Boris's first comment replying to my initial post makes more sense to me now. It's going to take a little while longer for me to work through the notes, but I'm busy this afternoon and on leave tomorrow and Monday, so I will come back to this next week.

[University of Olso notes](https://www.uio.no/studier/emner/matnat/ifi/nedlagte-emner/INF-MAT5340/v05/undervisningsmateriale/kap1.pdf)

**Reply by borisburger** [reply]
Thank you for the link to the University of Oslo lecture notes, the material definitely contains a lot more exposition and detail that the Wikipedia articles lack - the Wikipedia articles are written very sparsely and are not really suitable as the first reading on this subject... I will download also the subsequent chapters on BSplines and approximation in the hope of getting some more insight how BSplines fit into the picture.

Can you please attach an image with your input dataset (coordinate points of input polygons) fitted with the spline that passes through all of them? I am interested especially in the density of the input dataset - for example if you have say a jetstream which is 2000km long or a turbulence area that covers an area the size of a country, how many points are on your polygons (or basically what is roughly the precision in kilometers of the polygon points) before they are interpolated with the spline... 

**aggdraw** - the docs are on https://aggdraw.readthedocs.io/en/latest/ and it should be possible to install it with "pip install aggdraw" if pip package manager is available in your Python distribution. I think it is used in the example just for the final plotting of the resulting Beziers. AGG is originally a C++ drawing API (https://en.wikipedia.org/wiki/Anti-Grain_Geometry), which became really popular and aggdraw is a Python "binding" on top of AGG. It is a remarkable and truly brilliant drawing library (we used it partially in our map drawing code). Unfortunately the original author of AGG passed away prematurely and the library is in sort of a maintenance mode right now with a community trying to keep it alive...

**Reply by GAndrsn** [reply]
I've attached a plot of some of my SigWx Object code output, which shows plotted hazard objects. I currently have small crosses added at the points of input data visible on CAT (yellow dashed lines) and JET (green arrows) objects. Typical point numbers will range from a half dozen to a dozen, but some large areas (e.g. CB areas around the tropics) can have several dozen points.

![SigWx_20210525_00Z_T+24_EUR_col](https://user-images.githubusercontent.com/32956809/119616781-e6674a00-bdf8-11eb-9ba2-9878575e593e.png)

JUst to flag up, I work 50:50 between aviation science (this work) and as a trainer at the Met Office. I am about to finish my current stint in science, and won't be back until mid-August. Does this issue have a specific deadline it needs resolved by? As no-one else here has the level of knowledge of the issue I have currently, and I won't be able to spend any time on it over the next couple of months.

**Reply by blchoy** [reply]
> JUst to flag up, I work 50:50 between aviation science (this work) and as a trainer at the Met Office. I am about to finish my current stint in science, and won't be back until mid-August. Does this issue have a specific deadline it needs resolved by? As no-one else here has the level of knowledge of the issue I have currently, and I won't be able to spend any time on it over the next couple of months.

That depends on whether we need to implement a get-around that involves a change in the IWXXM schema.  For your information we are going to submit this version of IWXXM for WMO approval in late June/early July.  Up to this moment I can visualize the issue to be confined to iwxxm:phenomenonGeometry and since we have not yet include any constraints on the elements to be used within its context the chance for requiring a schema change is low.  Having said that, as we are also including [an example](http://schemas.wmo.int/iwxxm/2021-2RC1/examples/WAFS-Example.xml) in the schema distribution, and vendors may start to develop their solutions once this version of IWXXM has approved, we may need to decide what we should do with the example before submission for approval before late June upon the progress of this investigation.

**Reply by blchoy** [reply]
> I've attached a plot of some of my SigWx Object code output, which shows plotted hazard objects. I currently have small crosses added at the points of input data visible on CAT (yellow dashed lines) and JET (green arrows) objects. Typical point numbers will range from a half dozen to a dozen, but some large areas (e.g. CB areas around the tropics) can have several dozen points.

I noticed that the CAT polygons are quite close over northern Spain.  Are you going to generate the tropopause contours with the same method?  We may also want to check for possibility of crossing contours.

**Reply by GAndrsn** [reply]
@blchoy I suspected that the change may be desired before late-June. From the solution I found yesterday I just need to know how to convert the individual Bezier segments into whatever the representation is for these is in IWXXM, presumably as a single curve. I'll need guidance from yourself or Boris on this though, the online information on this is very thin.

> I noticed that the CAT polygons are quite close over northern Spain. Are you going to generate the tropopause contours with the same method? We may also want to check for possibility of crossing contours.

Overlapping of CAT/TURB hazards of different severity levels is still an issue to be resolved (in this case, the smaller areas are severe CAT, the outer area is moderate). I already have a solution in place for preventing TROP contours overlapping, that works in the majority of cases (but not all).

I'm going to move my comments from yesterday into replies in this thread, as I accidentally added them is new threads.

**Reply by GAndrsn** [reply]
I've tried working through the University of Oslo notes and everything up to Section 1.5 seems reasonable and helpful, but from 1.5 on I get stuck and lose what is being discussed. I've had a go at just experimenting with code instead.

I've found that the Python SciPy interpolation functions allow you to calculate the derivative of a spline as well as the points, so I had a go at applying the conversion to a Bezier curve mentioned by Boris in his comment above from 13th May (adding 1/3 of the derivative vectors to get P1 and P2 points). I did this for a spline interpolating 5 points in a square, with the spline set to be periodic (see plot below). The green line is the spline interpolated using the scipy.interpolate package.

![square_interpolation](https://user-images.githubusercontent.com/32956809/119782433-192a4480-bec4-11eb-8c28-731102b0a71f.png)

The derivative at point 0 is [-3, 3], so I tried a Bezier curve with points at [0, 0], [-1, 1], [-1, 0] and [0, 1], but this didn't fit. After a bit of experimentation though, I found that using points at [-0.25, 0.25] and [-0.25, 0.75] does fit. I think this might be because the output of the spline function uses knots at a regular spacing of 0.25 (full knot vector [-0.75 -0.5 -0.25 0. 0.25 0.5 0.75 1. 1.25 1.5 1.75]). I'll try this with other shapes.

![bezier_conversion](https://user-images.githubusercontent.com/32956809/119782448-1e878f00-bec4-11eb-977d-cf238828736b.png)

A couple of points to note here. I might have something that works, but don't really know what I'm doing! Also, I might end up with an output of a whole bunch of 4-point cubic Bezier curves, but I don't know how these would be stitched together for GML Bezier curve type output.

**Reply by GAndrsn** [reply]
So I've expanded the work I mentioned in thelast comment to work for more points and more general shapes, and it looks like I can convert a cubic B-spline generated using the Python scipt.interpolate library with n points into n-1 cubic Bezier curve segments, each with 4 control points. Presumably there is a way to convert these to a single piecewise Bexier curve representation for IWXXM/GML representation. As I've mentioned, however, I still don't 100% understand how I've got to this!

The Bezier cuvers in the plot below are yellow; the spline is blue; each are set to 50% transparency, so when plotted together, the output line will be green if the lines overlap exactly.

![bezier_conversion_2](https://user-images.githubusercontent.com/32956809/119782540-38c16d00-bec4-11eb-920b-f408c465faba.png)

**Reply by borisburger** [reply]
@GAndrsn thank you for the examples. Graeme, what really confuses me right now is that we started the discussion with **gml:CubicSpline**, but later on it turned out that you are using SciPy to interpolate with **B-Splines**, which in GML would have to be represented using **gml:BSpline** element instead. Normally the B-Spline would not interpolate the points on input, but your spline somehow does, which is really baffling, but maybe I just need to read up how exactly SciPy does this.

Normally the B-Spline would have to be converted to Bezier using the Boehm knot insertion algorithm. I think the trick with calculating derivatives of the B-Spline and generating Bezier control points will not work when B-Spline is on input, but really it is the knot insertion algorithm that is required. However, if your solution works and produces identical output to the B-Spline, than perhaps your solution is right and we just need to think how to produce the final Bezier spline control points out of it...

If the final decision was to use gml:CubicSpline (the C2 continuous one), I have even less clue how to convert the B-Spline into C2 gml:CubicSpline, than to a Bezier (where at least I understand the knot insertion algorithm is the way to go).

Are you able to spare some time for a short call, whenever suitable for you? For example using Google Meet or MS Teams, or Cisco Webex, so that we could also share the screen and discuss. I am in CET time zone (GMT+2)...

**Reply by blchoy** [reply]
I think we do not need to limit ourselves to gml:CubicSpline and gml:BSpline but other interpolation schemes that can be described by GML, as long as it is the kind of smooth curve that is both fit-for-purpose (meeting WAFCs' expectation) and achievable (something that developers can do).

In case there is an agreed discussion time I can arrange a Webex meeting.  I normally available 12-16UTC during weekdays (but not for today 28 May, as I have another discussion with Anna).

**Reply by GAndrsn** [reply]
@borisburger I don't have an in-depth understanding of the differences between types of spline. My assumption has been that the scipy.interpolate functions produce a B-Spline, which is a type of spline, and that a third-order B-Spline could be described as a cubic spline. As such when we've represented the data in IWXXM we've used the cubic spline type, but that has been based on advice from Choy. As a meteorologist the details here really aren't my area of expertise!

By default, the B-Splines produced by the scipy functions don't interpolate the input points, but the argument s (a "smoothing condition") can be set to 0, in which case it does ([see details here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.splprep.html)). There are references at the bottom of the page by P. Dierckx that might clarify what's going on, but they require payment to access, and I doubt I would understand them anyway. 

I am happy to join a Webex next week, Tues-Thurs (I am not in next Mon or Fri). My only meetings so far are Tues and Weds, 13-14UTC, but any other time is fine.

**Reply by borisburger** [reply]
I am really sorry for being slow to respond, this week is a bit crazy for me due to a milestone presentation we are preparing for a customer. I am available:
- Wednesday 07-13 UTC and 14-16 UTC.
- Thursday 12-16 UTC (at least I hope the presentation we will be doing for the customer will be over by 12UTC)

**Reply by blchoy** [reply]
So the commonly available slots are Wed 14-16UTC and Thu 13-16UTC.

May I suggest we try Wed 1430-1600UTC first?  I will send a Webex invitation now.

**Reply by borisburger** [reply]
I confirm my availability Wed 14:30-1600 UTC.

**Comment by blchoy** [comment]
Thank you for Graeme and Boris joining the discussion yesterday.  I would like to highlight the outcomes and follow up actions below.  Feel free to comment or modify.

1. The interpolation function in SciPy seems to be able to comply with the C2 requirement of a natural cubic spline which gml:CubicSpline is referring.  A number of documents were referenced during the discussion and if Boris could drop a few lines what they are Choy may be able to invite some of his colleagues to confirm.
2. It may be worth trying to reproduce the curves Graeme produced with a different software to confirm this could be done.  Choy will see if there is any volunteer to conduct a test.
3. The existing WAFC's guidance on rendering may need to beef up to include necessary details to facilitate reproduction of the curves.  Graeme will work with Karen on this afterwards.
4. There was a discussion on other possible issues, including "dateline crossing features" and "features close to the north or south pole".  It appeared that as individual object could specify its own CRS it may help to better describe their geometry.  As this would not affect the IWXXM schemas, Choy and Graeme would re-visit this closer to the time when WAFCs is going releasing data for operation trial.

Finally there is hope for us to safely close this topic.  :)

I would also like to forward Boris' comment that what Graeme has been doing with regard to the generation of weather objects is very impressive.  I can't agree more with him.

**Comment by GAndrsn** [comment]
Thank you to Choy for re-opening this discussion, so Boris can potentially contribute as well. I emailed Choy with the following question, which I hope one of you can answer: 

> Can you clarify for me exactly what the end vectors should represent? At the moment I think I’ve been calculating them as unit vectors in the direction of the start/end of the spline, but I suspect this is incorrect. I have a vague recollection of mention of these vectors representing derivatives of the spline. Could you clarify this for me? I would think that whether the end vectors are correct or incorrect will rapidly become apparent if we try reading the IWXXM data into software that can interpret glm:CubicSpline data.

**Reply by borisburger** [reply]
I am sorry for not continuing this discussion. The two vectors should really represent the derivatives at start and end. I am attaching an excerpt from **ISO 19107:2019 Geographic Information - Spatial Schema**:

![image](https://user-images.githubusercontent.com/21176728/129014522-6328ecda-5db0-42aa-a324-347f968061bb.png)

**Reply by GAndrsn** [reply]
Great, thanks for the quick response, I will look at how I implement this in my IWXXM writing code.

**Reply by GAndrsn** [reply]
Using the scipy.interpolate.splev() function with the 'der' keyword set to 1 appears to create a first derivative vector that gives the desired tangent vector at any point along the spline. See output below for interpolation of curve y = -(0.2*(x-34.5))**2 between 0 < x < 70:

![scipy_spline_derivatives](https://user-images.githubusercontent.com/32956809/129357472-6f777080-5a10-43ae-8da8-78fbae308e47.png)

Having done a bit more testing, I believe that the magnitude of the derivative vectors is caused by the derivative vector representing (dx/ds, dy/ds), where s is the parameter that varies from 0 to 1 along the length of the spline. As such, for a quadratic curve from 0 to 70, ds will be less than 1/100th of the length of the spline per unit length of the spline, and so the magnitude of the derivative vector will be greater than 100. How this is scaled will need to be considered for the implementation of this type of derivative for lat/lon data.

I can apply this new approach to calculating end vectors to my SigWx IWXXM code, but I won't know I've got it right until the data is read into a proper visualisation software package and I can see how the data is interpreted.
